src/esphome/sensor/sds011_component.cpp

```diff
@@ -0,0 +1,258 @@
+    } else {
+      // next byte
+      this->data_index_++;
+    }
+  }
+}
+
+void SDS011Component::update() {
+  if (!this->get_query_mode()) {
+    return;
+  }
+  this->force_query_ = true;
+}
+
+SDS011Sensor *SDS011Component::make_pm_2_5_sensor(const std::string &name) {
+  return this->pm_2_5_sensor_ = new SDS011Sensor(name);
+}
+
+SDS011Sensor *SDS011Component::make_pm_10_0_sensor(const std::string &name) {
+  return this->pm_10_0_sensor_ = new SDS011Sensor(name);
+}
+
+float SDS011Component::get_setup_priority() const {
+  return setup_priority::HARDWARE_LATE;
+}
+
+void SDS011Component::set_query_mode(bool query_mode) {
+  this->query_mode_ = query_mode;
+}
+
+void SDS011Component::set_rx_mode_only(bool rx_mode_only) {
+  this->rx_mode_only_ = rx_mode_only;
+}
+
+void SDS011Component::set_update_interval(uint32_t update_interval) {
+  this->update_interval_ = update_interval;
+}
+
+bool SDS011Component::get_query_mode() const {
+  return this->query_mode_;
+}
+
+bool SDS011Component::get_rx_mode_only() const {
+  return this->rx_mode_only_;
+}
+
+uint32_t SDS011Component::get_update_interval() const {
+  return this->update_interval_;
+}
+
+void SDS011Component::sds011_write_command_(const uint8_t *command_data) {
+  this->flush();
+  this->write_byte(SDS011_MSG_HEAD);
+  this->write_byte(SDS011_COMMAND_ID_REQUEST);
+  this->write_array(command_data, SDS011_DATA_REQUEST_LENGTH);
+  this->write_byte(sds011_checksum_(command_data, SDS011_DATA_REQUEST_LENGTH));
+  this->write_byte(SDS011_MSG_TAIL);
+}
+
+uint8_t SDS011Component::sds011_checksum_(const uint8_t *command_data, uint8_t length) const {
+  uint8_t sum = 0;
+  for (uint8_t i = 0; i < length; i++) {
+    sum += command_data[i];
+  }
+  return sum;
+}
+
+optional<bool> SDS011Component::check_byte_() const {
+  uint8_t index = this->data_index_;
+  uint8_t byte = this->data_[index];
+
+  if (index == 0) {
+    return byte == SDS011_MSG_HEAD;
+  }
+
+  if (index == 1) {
+    return byte == SDS011_COMMAND_ID_DATA;
+  }
+
+  if ((index >= 2) && (index <= 7)) {
+    return true;
+  }
+
+  if (index == 8) {
+    // checksum is without checksum bytes
+    uint8_t checksum = sds011_checksum_(this->data_ + 2, SDS011_DATA_RESPONSE_LENGTH);
+    if (checksum != byte) {
+      ESP_LOGW(TAG, "SDS011 Checksum doesn't match: 0x%02X!=0x%02X", byte, checksum);
+      return false;
+    }
+    return true;
+  }
+
+  if (index == 9) {
+    if (byte != SDS011_MSG_TAIL) {
+      return false;
+    }
+  }
+
+  return {};
+}
+
+void SDS011Component::parse_data_() {
+  this->status_clear_warning();
+  const float pm_2_5_concentration = this->get_16_bit_uint_(2) / 10.0f;
+  const float pm_10_0_concentration = this->get_16_bit_uint_(4) / 10.0f;
+  const uint16_t device_id = this->get_16_bit_uint_(6);
+
+  ESP_LOGD(TAG,
+           "Got PM2.5 Concentration: %.1f µg/m³, PM10.0 Concentration: %.1f "
+           "µg/m³ from device 0x%04x",
+           pm_2_5_concentration, pm_10_0_concentration, device_id);
```

Comment:
Can there be more than one device connected to one bus? If so, we should split add a device_id option so that users can set up multiple instances of this class and select which sensor to read from.

Reply:
It is not directly mentioned that multiple sensors are possible. However, since the data frames in the protocol contain a device id and a broadcast id (`0xffff`) exists, one would think that this is possible. But for me is not clear how possible collision will be handled on the bus. I have only one device and can not test it.

IMHO, to support multiple sensors on the bus, we need to split the component into two parts (similar to the dallas component):
1. Hub component to configure the uart port and handle incoming frames.
2. Sensor component with device_id (default to broadcast id) .

I had this in mind (and basically like that idea), but thought it was too over-engineered, especially since it is not clear if multiple sensors are possible.

Reply:
Ok, I guess especially since we support SW UART on the ESP8266s there's not a real reason to support more than one device per bus. Even if we could, the sensors could run into trouble when they both hold the bus high/low at the same time.

Reply:
Let's just remove that from the logs then, since the normal user shouldn't be concerned with this.

Reply:
Done